src/MyInfoGovClient.class.ts

```diff
@@ -1,609 +1,331 @@
+import qs from 'qs'
 import crypto from 'crypto'
-import jose from 'node-jose'
-import path from 'path'
 import axios from 'axios'
-import qs from 'qs'
-import { IPersonBasic, MyInfoAttribute } from './myinfo-types'
+import { hasProp, objToSearchParams, sortObjKeys } from './util'
+import { verify as verifyJwt } from 'jsonwebtoken'
+import { IPerson } from './myinfo-types'
 
-export enum Mode {
+/**
+ * Mode in which to initialise the client, which determines the
+ * MyInfo endpoint to call.
+ */
+export enum MyInfoMode {
   Dev = 'dev',
   Staging = 'stg',
-  Production = 'prod'
-}
-
-export interface IConfig {
-  realm: string
-  appId: string
-  singpassEserviceId: string
-  privateKey: Buffer | string
-  clientId?: string
-  mode?: Mode
+  Production = 'prod',
 }
 
-export interface IPersonBasicRequest {
-  uinFin: string,
-  requestedAttributes?: string[],
-  txnNo?: string,
-  singpassEserviceId?: string
+const BASE_URL: { [M in MyInfoMode]: string } = {
+  [MyInfoMode.Dev]: 'http://localhost:5156/myinfo/v3',
+  [MyInfoMode.Staging]: 'https://myinfosgstg.api.gov.sg/gov/test/v3',
+  [MyInfoMode.Production]: 'https://myinfosg.api.gov.sg/gov/v3',
 }
 
-interface IBaseStringSpec {
-  httpMethod: string
-  url: string
-  appId: string
+/**
+ * Parameters for MyInfoGovClient constructor.
+ */
+export interface IMyInfoConfig {
   clientId: string
+  clientSecret: string | Buffer
   singpassEserviceId: string
-  nonce: string
-  requestedAttributes: string[]
-  timestamp: number
-  txnNo?: string
+  redirectEndpoint: string
+  clientPrivateKey: string | Buffer
+  myInfoPublicKey: string | Buffer
+  mode?: MyInfoMode
 }
 
-type MyInfoParsedResponse = Omit<IPersonBasic, 'uinFin'>
-type MyInfoResponse = MyInfoParsedResponse | string
-
-const BASE_URL: { [M in Mode]: string } = {
-  [Mode.Dev]: 'https://myinfosgstg.api.gov.sg/gov/dev/v1/',
-  [Mode.Staging]: 'https://myinfosgstg.api.gov.sg/gov/test/v2/',
-  [Mode.Production]: 'https://myinfosg.api.gov.sg/gov/v2/',
+/**
+ * Parameters to create a redirect URL to initialise MyInfo login.
+ */
+export interface IAuthRequest {
+  purpose: string
+  requestedAttributes: string[]
```

Comment:
it would be good to encode the list of supported MyInfo attributes here

Reply:
Yeah we could make this argument of type `MyInfoAttribute[]`, but then we wouldn't be able to support libraries (eg Forms) which only support a subset of attributes. Here's a playground example which demonstrates what I mean:

![Screenshot 2021-02-08 at 11 18 13 AM](https://user-images.githubusercontent.com/29480346/107172059-5941e600-69ff-11eb-9e58-45e54018fc69.png)

In other words, we'd force Forms to cast its own `MyInfoAttribute` enum (which has a subset of the attributes) to be the type of the MyInfoGovClient `MyInfoAttribute` (which has all the attributes).

Any ideas for a workaround? Or should we just force upstream libraries to do the casting?

Reply:
nevermind, figured it out

![Screenshot 2021-02-08 at 11 38 13 AM](https://user-images.githubusercontent.com/29480346/107173169-23523100-6a02-11eb-96c3-67fd8a9770ab.png)